### ISSUE TITLE
How to optimize several variables concurrently？

### ISSUE THREAD

**Issue by AlexLiuyuren** [hyperopt/hyperopt#394]
Assume that I want to optimize a Sphere function, which looks like:
```python
def sphere(x):
    """
    x is a numpy vector. e.g. x = np.array([0.5, 0.6])
    """
    return sum(x ** 2)
```
Apparently, the optimal position is [0, 0, ....]. Assume that the dimension size is 3, which means we have three variable to optimize, and the search space is [-1, 1] for each dimension, how to optimize the function? I've read through the tutorial and find that all examples optimize a function with one variable. Is that possible to  optimize several variables concurrently？Thank you!

**Comment by benfred** [comment]
There might be a better way to do this, but one thing that works is code like:

```python
from hyperopt import fmin, tpe, hp,
import numpy as np

def sphere(x):
    return sum(np.array(x) ** 2)

fmin(sphere, space=[hp.uniform(str(dim), -1, 1) for dim in range(3)], algo=tpe.suggest, max_evals=200)
```

**Comment by AlexLiuyuren** [comment]
Thank you! This works perfectly.

**Comment by ashutoshnayakIE** [comment]
Hi,
I am trying a similar method where I have a 2D vector of variables, but it is not taking 2D vectors as space. If I convert these 2d vectors into a 1D vector and then when I convert it using reshape, it does not work. Let me know how can I do that. I have a matrix form for the variables.

**Comment by bjkomer** [comment]
The output of your objective function must be a scalar (i.e. a single floating point value). The input can be anything. If your function has multiple outputs you want to minimize, you need to combine them in some way. A simple way to do that is to take the norm, which for a vector will just be the sum of squares of the output (like the example above). Here is a general way to do this:

```python
def objective(params):
    loss = my_func_that_returns_a_matrix(params)
    # use matrix or vector norm to create a scalar
    return np.linalg.norm(loss)
```

for your example you would simply return `np.linalg.norm(x)`

**Comment by CoinCheung** [comment]
@bjkomer Hi, Do you know how to optimizer three float numbers with sum to be 1? It is like this:
```
space=[hp.uniform('w1', 0, 1), hp.uniform('w2', 0, 1), hp.uniform('w3', 0, 1), ]
```
The problem is that I hope `w1`, `w2`, and `w3` to satisfy the relation that `w1+w2+w3=1`. How could I add this condition ?

**Comment by bjkomer** [comment]
@CoinCheung there isn't a direct way to add constraints like that, but one way to achieve similar results is to treat those parameters as relative weightings, and just normalize the result to sum to one before using them in your objective function (i.e. instead of `w1` in your function, just use `w1/(w1+w2+w3)`).
